src/tm_mad/lib/ceph.rb

```diff
@@ -18,39 +18,270 @@
+
+                    cmds = d.backup_cmds(backup_dir, ds, live)
+                    snap_cmd << cmds[:snapshot]
+                    expo_cmd << cmds[:export]
+                    clup_cmd << cmds[:cleanup]
+                end
+
+                freeze, thaw =
+                    if live
+                        fsfreeze(@xml, deploy_id)
+                    else
+                        ['', '']
+                    end
+
+                <<~EOS
+                    set -ex -o pipefail
+
+                    # ----------------------
+                    # Prepare backup folder
+                    # ----------------------
+                    [ -d #{backup_dir} ] && rm -rf #{backup_dir}
+
+                    mkdir -p #{backup_dir}
+
+                    echo "#{Base64.encode64(@xml)}" > #{backup_dir}/vm.xml
+
+                    # --------------------------------
+                    # Create Ceph snapshots for disks
+                    # --------------------------------
+                    #{freeze}
+
+                    #{snap_cmd}
+
+                    #{thaw}
+
+                    # --------------------------
+                    # export, convert & cleanup
+                    # --------------------------
+                    #{expo_cmd}
+
+                    #{clup_cmd}
+                EOS
+            end
+
+        end
+
         # Ceph disks
         class Disk
 
+            # DON'T CHANGE THIS CONSTANT; will break existing incremental backups
+            INC_SNAP_PREFIX  = 'one_backup_'
+
             attr_reader :id, :vmid, :source, :clone, :rbd_image, :rbd_cmd
 
-            # @param vmid [Integer]
+            # @param vm_xml [String, REXML::Element]
             # @param disk_xml [String, REXML::Document, REXML::Element]
             # @return [Disk]
-            def initialize(vmid, disk_xml)
+            def initialize(vm_xml, disk_xml)
+                vm_xml = REXML::Document.new(vm_xml) if vm_xml.is_a?(String)
                 disk_xml = REXML::Document.new(disk_xml) if disk_xml.is_a?(String)
 
-                @id     = disk_xml.elements['DISK_ID'].text.to_i
-                @vmid   = vmid
-                @source = disk_xml.elements['SOURCE'].text
-                @clone  = disk_xml.elements['CLONE'].text == 'YES'
+                @vm   = vm_xml
+                @vmid = @vm.elements['TEMPLATE/VMID'].text
+                @id   = disk_xml.elements['DISK_ID'].text.to_i
+                @type = disk_xml.elements['TYPE'].text
+                @pool = disk_xml.elements['POOL_NAME'].text
 
-                @rbd_image =
-                    if @clone
-                        "#{@source}-#{@vmid}-#{@id}"
-                    else
-                        @source
-                    end
+                if volatile?
+                    @source      = nil
+                    @clone       = nil
+                    @rbd_image   = "#{@pool}/one-sys-#{@vmid}-#{id}"
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

A cross-site scripting (XSS) vulnerability was detected. User-controlled data is being output without proper sanitization, allowing an attacker to inject malicious scripts. This could lead to session hijacking, malware installation, or phishing attacks. Always sanitize and encode user inputs before including them in the output, using context-appropriate encoding methods. Learn More - https://cwe.mitre.org/data/definitions/79.html.